repos/system_upgrade/common/actors/cloud/checkrhui/libraries/checkrhui.py

```diff
@@ -0,0 +1,221 @@
+from leapp.libraries.stdlib import api
+from leapp.models import (
+    CopyFile,
+    DNFPluginTask,
+    InstalledRPM,
+    RHUIInfo,
+    RpmTransactionTasks,
+    TargetRHUIPostInstallTasks,
+    TargetRHUIPreInstallTasks,
+    TargetRHUISetupInfo,
+    TargetUserSpacePreupgradeTasks
+)
+
+
+MatchingSetup = namedtuple('MatchingSetup', ['name', 'description'])
+
+
+def into_set(pkgs):
+    if isinstance(pkgs, set):
+        return pkgs
+    if isinstance(pkgs, str):
+        return {pkgs}
+    return set(pkgs)
+
+
+def find_rhui_setup_matching_src_system(installed_pkgs, rhui_map):
+    src_ver = version.get_source_major_version()
+    arch = api.current_actor().configuration.architecture
+
+    matching_setups = []
+    for rhui_variant, known_variant_setups in rhui_map.items():
+        for setup in known_variant_setups:
+            if setup.os_version != src_ver or setup.arch != arch:
+                continue
+            if setup.clients.issubset(installed_pkgs):
+                matching_setups.append(MatchingSetup(name=rhui_variant, description=setup))
+
+    if not matching_setups:
+        return None
+
+    # In case that a RHUI variant uses a combination of clients identify the maximal client set
+    matching_setups_by_size = sorted(matching_setups, key=lambda match: -len(match.description.clients))
+
+    match = matching_setups_by_size[0]  # Matching setup with the highest number of clients
+    if len(matching_setups) == 1:
+        return match
+
+    if len(matching_setups_by_size[0].description.clients) == len(matching_setups_by_size[1].description.clients):
+        msg = 'Could not identify source RHUI setup (ambiguous setup)'
+        details = ('Installed client identify the provider \'{match0.name}\' (identified by clients '
+                   '{match0.description.clients} and also the {match1.name} (identified by clients '
+                   '{match1.description.clients})')
+        details.format(match0=matching_setups_by_size[0], match1=matching_setups_by_size[1])
+        raise StopActorExecutionError(message=msg, details={'details': details})
+
+    return match
+
+
+def determine_target_setup_desc(cloud_map, variant):
+    variant_setups = cloud_map[variant]
+    target_major = version.get_target_major_version()
+    arch = api.current_actor().configuration.architecture
+
+    for setup in variant_setups:
+        if setup.arch == arch and setup.os_version == target_major:
+            return setup
+    return None
+
+
+def inhibit_if_leapp_pkg_to_access_target_missing(installed_pkgs, provider, target_setup_desc):
+    pkg_name = target_setup_desc.leapp_pkg
+    if pkg_name not in installed_pkgs:
+        reporting.create_report([
+            reporting.Title('Package "{}" is missing'.format(pkg_name)),
+            reporting.Summary(
+                'On {} using RHUI infrastructure, a package "{}" is needed for '
+                'in-place upgrade'.format(provider.upper(), pkg_name)
+            ),
```

Comment:
I agree that the 'provider' is an imprecise and confusing name. I've changed it to variant. Since cloud providers distinguish
RHEL and RHEL SAP in their offerings, I believe it should be OK to present the user with the internal _variant_ since it closely matches the thing the user sees when spawning a VM.

Reply:
Should be fixed.